File: typescript/src/storage/ipfsStorageProvider.ts
/**
 * IPFSStorageProvider — IPFS pinning service storage.
 *
 * Uses an HTTP pinning API (Pinata-compatible `pinJSONToIPFS`) for upload
 * and an IPFS gateway for download/exists. Uses the global `fetch` (Node
 * 20+ has it built in) rather than a dependency, since the TS SDK has no
 * `httpx` equivalent to reach for.
 *
 * Port of `python/bnbagent/storage/ipfs_storage_provider.py`.
 */

import { getEnv } from "../core/envUtil.js";
import { StorageError } from "../errors.js";
import { StorageProvider } from "./storageProvider.js";

const DEFAULT_PINNING_URL = "https://api.pinata.cloud/pinning/pinJSONToIPFS";
const DEFAULT_GATEWAY_URL = "https://gateway.pinata.cloud/ipfs/";

// CIDv0: Qm + 44 base58 chars; CIDv1: b + base32 (58+ chars)
const CID_RE = /^(Qm[1-9A-HJ-NP-Za-km-z]{44}|b[a-z2-7]{58,})$/;

/**
 * IPFS storage via HTTP pinning API.
 *
 * @param pinningApiUrl - e.g. "https://api.pinata.cloud/pinning/pinJSONToIPFS"
 * @param pinningApiKey - Bearer token (JWT) for the pinning service
 * @param gatewayUrl - e.g. "https://gateway.pinata.cloud/ipfs/"
 */
export class IPFSStorageProvider extends StorageProvider {
  private readonly pinningUrl: string;
  private readonly apiKey: string;
  private readonly gateway: string;

  constructor(
    pinningApiUrl: string,
    pinningApiKey: string,
    gatewayUrl: string = DEFAULT_GATEWAY_URL,
  ) {
    super();
    this.pinningUrl = pinningApiUrl;
    this.apiKey = pinningApiKey;
    this.gateway = gatewayUrl.replace(/\/+$/, "");
  }

  static fromEnv(): IPFSStorageProvider {
    const apiKey = getEnv("STORAGE_API_KEY");
    if (!apiKey) {
      throw new Error("STORAGE_API_KEY required for IPFSStorageProvider");
    }
    return new IPFSStorageProvider(
      getEnv("STORAGE_API_URL", DEFAULT_PINNING_URL) as string,
      apiKey,
      getEnv("STORAGE_GATEWAY_URL", DEFAULT_GATEWAY_URL) as string,
    );
  }

  async upload(
    data: Record<string, unknown>,
    filename?: string,
  ): Promise<string> {
    const headers = {
      Authorization: `Bearer ${this.apiKey}`,
      "Content-Type": "application/json",
    };

    let pinName: string;
    if (filename) {
      pinName = filename.replaceAll(".json", "");
    } else {
      const jobId = IPFSStorageProvider.extractJobId(data);
      pinName = jobId ? `erc8183-job-${jobId}` : "deliverable";
    }

    const payload = {
      pinataContent: data,
      pinataMetadata: { name: pinName },
    };

    const resp = await fetch(this.pinningUrl, {
      method: "POST",
      headers,
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(60_000),
    });
    if (!resp.ok) {
      throw new StorageError(
        `Pinning request failed: ${resp.status} ${resp.statusText}`,
      );
    }
    const result = (await resp.json()) as Record<string, unknown>;
    const cid =
      (result.IpfsHash as string | undefined) ??
      (result.cid as string | undefined);
    if (!cid) {
      throw new StorageError(
        `Unexpected pinning response: ${JSON.stringify(result)}`,
      );
    }

    const ipfsUrl = `ipfs://${cid}`;
    return ipfsUrl;
  }

  async download(url: string): Promise<Record<string, unknown>> {
    const cid = IPFSStorageProvider.extractCid(url);
    const gatewayUrl = `${this.gateway}/${cid}`;

    const resp = await fetch(gatewayUrl, {
      signal: AbortSignal.timeout(30_000),
    });
    if (!resp.ok) {
      throw new StorageError(
        `Gateway request failed: ${resp.status} ${resp.statusText}`,
      );
    }
    return (await resp.json()) as Record<string, unknown>;
  }

  async exists(url: string): Promise<boolean> {
    const cid = IPFSStorageProvider.extractCid(url);
    const gatewayUrl = `${this.gateway}/${cid}`;

    try {
      const resp = await fetch(gatewayUrl, {
        method: "HEAD",
        signal: AbortSignal.timeout(10_000),
      });
      return resp.status === 200;
    } catch {
      return false;
    }
  }

  /** Convert an `ipfs://` URL to an HTTP gateway URL for browser access. */
  getGatewayUrl(ipfsUrl: string): string {
    const cid = IPFSStorageProvider.extractCid(ipfsUrl);
    return `${this.gateway}/${cid}`;
  }

  static extractCid(url: string): string {
    let cid: string;
    if (url.startsWith("ipfs://")) {
      cid = url.slice(7);
    } else if (url.includes("/ipfs/")) {
      cid = url.split("/ipfs/").pop() as string;
    } else {
      cid = url;
    }
    cid = cid.replace(/^\/+|\/+$/g, "");
    if (!CID_RE.test(cid)) {
      throw new StorageError(`Invalid IPFS CID format: ${cid}`);
    }
    return cid;
  }

  private static extractJobId(data: Record<string, unknown>): unknown {
    const job = data.job;
    if (job && typeof job === "object" && !Array.isArray(job)) {
      return (job as Record<string, unknown>).id;
    }
    return undefined;
  }
}

File: python/bnbagent/storage/ipfs_storage_provider.py
"""
IPFSStorageProvider — IPFS pinning service storage.

Uses HTTP API (Pinata/Infura/Web3.Storage) for upload and an IPFS gateway for download.
Requires `httpx` (optional dependency).
"""

from __future__ import annotations

import json
import logging
import os
import re

import httpx

from ..exceptions import StorageError
from .storage_provider import StorageProvider

logger = logging.getLogger(__name__)

DEFAULT_MAX_IPFS_UPLOAD_BYTES = 8 * 1024 * 1024


def _max_upload_bytes_from_env() -> int:
    raw = os.getenv("STORAGE_MAX_UPLOAD_BYTES")
    if raw is None:
        return DEFAULT_MAX_IPFS_UPLOAD_BYTES
    try:
        value = int(raw)
        if value <= 0:
            raise ValueError
        return value
    except ValueError:
        logger.warning(
            "STORAGE_MAX_UPLOAD_BYTES=%r invalid, using %d",
            raw,
            DEFAULT_MAX_IPFS_UPLOAD_BYTES,
        )
        return DEFAULT_MAX_IPFS_UPLOAD_BYTES


class IPFSStorageProvider(StorageProvider):
    """
    IPFS storage via HTTP pinning API.

    Args:
        pinning_api_url: e.g. "https://api.pinata.cloud/pinning/pinJSONToIPFS"
        pinning_api_key: Bearer token (JWT) for the pinning service
        gateway_url: e.g. "https://gateway.pinata.cloud/ipfs/"
        max_upload_bytes: Serialized pin request ceiling (default 8 MiB).
    """

    def __init__(
        self,
        pinning_api_url: str,
        pinning_api_key: str,
        gateway_url: str = "https://gateway.pinata.cloud/ipfs/",
        max_upload_bytes: int = DEFAULT_MAX_IPFS_UPLOAD_BYTES,
    ):
        if max_upload_bytes <= 0:
            raise ValueError("max_upload_bytes must be > 0")
        self._pinning_url = pinning_api_url
        self._api_key = pinning_api_key
        self._gateway = gateway_url.rstrip("/")
        self._max_upload_bytes = max_upload_bytes

    @classmethod
    def from_env(cls) -> IPFSStorageProvider:
        api_key = os.getenv("STORAGE_API_KEY")
        if not api_key:
            raise ValueError("STORAGE_API_KEY required for IPFSStorageProvider")
        return cls(
            pinning_api_url=os.getenv("STORAGE_API_URL")
            or "https://api.pinata.cloud/pinning/pinJSONToIPFS",
            pinning_api_key=api_key,
            gateway_url=os.getenv("STORAGE_GATEWAY_URL") or "https://gateway.pinata.cloud/ipfs/",
            max_upload_bytes=_max_upload_bytes_from_env(),
        )

    async def upload(self, data: dict, filename: str | None = None) -> str:
        headers = {
            "Authorization": f"Bearer {self._api_key}",
            "Content-Type": "application/json",
        }

        # Use provided filename or extract from job.id
        if filename:
            pin_name = filename.replace(".json", "")
        else:
            job_data = data.get("job", {})
            job_id = job_data.get("id") if isinstance(job_data, dict) else None
            pin_name = f"erc8183-job-{job_id}" if job_id else "deliverable"

        payload = {
            "pinataContent": data,
            "pinataMetadata": {"name": pin_name},
        }
        encoded = json.dumps(payload, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
        if len(encoded) > self._max_upload_bytes:
            raise StorageError(
                f"IPFS upload payload is {len(encoded)} bytes; "
                f"limit is {self._max_upload_bytes} bytes"
            )

        async with httpx.AsyncClient(timeout=60) as client:
            resp = await client.post(self._pinning_url, content=encoded, headers=headers)
            resp.raise_for_status()
            result = resp.json()

        cid = result.get("IpfsHash") or result.get("cid")
        if not cid:
            raise StorageError(f"Unexpected pinning response: {result}")

        ipfs_url = f"ipfs://{cid}"
        logger.info(f"[IPFSStorageProvider] Uploaded {pin_name} to {ipfs_url}")
        return ipfs_url

    async def download(self, url: str) -> dict:
        cid = self._extract_cid(url)
        gateway_url = f"{self._gateway}/{cid}"

        async with httpx.AsyncClient(timeout=30) as client:
            resp = await client.get(gateway_url)
            resp.raise_for_status()
            return resp.json()

    async def exists(self, url: str) -> bool:
        cid = self._extract_cid(url)
        gateway_url = f"{self._gateway}/{cid}"

        async with httpx.AsyncClient(timeout=10) as client:
            try:
                resp = await client.head(gateway_url)
                return resp.status_code == 200
            except httpx.HTTPError:
                return False

    def get_gateway_url(self, ipfs_url: str) -> str:
        """Convert ipfs:// URL to HTTP gateway URL for browser access."""
        cid = self._extract_cid(ipfs_url)
        return f"{self._gateway}/{cid}"

    # CIDv0: Qm + 44 base58 chars; CIDv1: b + base32 (58+ chars)
    _CID_RE = re.compile(r"^(Qm[1-9A-HJ-NP-Za-km-z]{44}|b[a-z2-7]{58,})$")

    @classmethod
    def _extract_cid(cls, url: str) -> str:
        if url.startswith("ipfs://"):
            cid = url[7:]
        elif "/ipfs/" in url:
            cid = url.split("/ipfs/")[-1]
        else:
            cid = url
        cid = cid.strip("/")
        if not cls._CID_RE.match(cid):
            raise StorageError(f"Invalid IPFS CID format: {cid}")
        return cid
